king_phisher/client/tabs/mail.py

```diff
@@ -866,6 +869,131 @@ def __init__(self, *args, **kwargs):
+		self.url_thread = None
+
+	@property
+	def _url_thread_is_ready(self):
+		return self.url_thread is None or not self.url_thread.is_alive()
+
+	def _url_custom_match_function(self, gtk_completion, key, tree_iter):
+		model = gtk_completion.get_model()
+		raw_iter_url = model[tree_iter][0]
+
+		if raw_iter_url.startswith(key):
+			return True
+		if not key.startswith('http'):
+			key = 'http://' + key
+
+		try:
+			key_parse = urllib.parse.urlparse(key)
+			parse_iter_url = urllib.parse.urlparse(raw_iter_url)
+		except ValueError:
+			self.logger.warning('value error when parsing uris for url entry completion')
+			return False
+
+		iter_hostname = parse_iter_url.netloc.partition(':')[0]
+		if key_parse.netloc:
+			if not iter_hostname.startswith(key_parse.netloc.split(':')[0]):
+				return False
+			if not parse_iter_url.path.startswith(key_parse.path):
+				return False
+
+		return True
+
+	def _url_completion_load(self, _):
+		self.config['server_config']['server.ssl'] = False
+		self.config['server_config']['server.addresses'] = self.application.rpc('config/get', 'server.addresses')
+		self.config['server_config']['server.vhost_directories'] = self.application.rpc('config/get', 'server.vhost_directories')
+		for address in self.config['server_config']['server.addresses']:
+			if address['ssl']:
+				self.config['server_config']['server.ssl'] = True
+		if not self._url_thread_is_ready:
+			return
+		self.url_thread = utilities.Thread(self._set_entry_completion_list_tsafe)
+		self.url_thread.start()
+
+	def _webserver_url_key_press_event(self, widget, event):
+		if event.type != Gdk.EventType.KEY_PRESS:
+			return
+		if not self._url_thread_is_ready:
+			return
+
+		keyval = event.get_keyval()[1]
+		if keyval == Gdk.KEY_F5:
+			self.url_thread = utilities.Thread(self._set_entry_completion_list_tsafe)
+			self.url_thread.start()
+		if (time.time() - self.url_completion_last_load_time) < self.refresh_frequency:
+			return
+		self.url_thread = utilities.Thread(self._set_entry_completion_list_tsafe)
+		self.url_thread.start()
+
+	def _get_url_templates(self):
+		return self.application.rpc.graphql("""\
+		{ 
+			siteTemplates 
+				{
+					edges 
+						{ 
+							node 
+								{ 
+									created 
+									hostname 
+									path 
+									metadata
+										{
+											authors
+											classifiers 
+											description 
+											pages
+										}
+								}
+						}
+				}
+		}""")
+
+	def _set_entry_completion_list_tsafe(self):
+		gui_utilities.glib_idle_add_once(self.url_list_store.clear)
+		url_information = self._get_url_templates()
+		if not url_information:
+			return
+		for edge in url_information['siteTemplates']['edges']:
+			for page in edge['node']['metadata']['pages']:
+				urls = self._build_urls(edge['node']['hostname'], page, edge['node']['path'])
+				if not urls:
+					continue
+				for url in urls:
+					gui_utilities.glib_idle_add_once(self.url_list_store.append, [url])
```

Comment:
So `Gtk.ListStore` doesn't have an `extend` method. I think this warrants a utility function to handle this in a single `glib_idle_add_*` call so we're not registering multiple one-shot routines to be done when it's idle. I'm surprised we haven't run into this elsewhere in the application yet.

I'm thinking something that would even handle the clear for you like this:
```python
# gui_utilities.py
def _store_extend(store, things):
    for thing in things:
        store.append(thing)

def glib_idle_add_store_extend(store, things, wait=False):
    """
    Extend a GTK store object (either :py:class:`Gtk.ListStore` or :py:class:`Gtk.TreeStore`) object using :py:func:`GLib.idle_add`. This function is suitable for use in non-main GUI threads for synchronizing data.

    :param store: The GTK storage object to add *things* to.
    :type store: :py:class:`Gtk.ListStore`, :py:class:`Gtk.TreeStore`
    :param tuple things: The array of things to add to *store*.
    :param bool wait: Whether or not to wait for the operation to complete before returning.
    """
    idle_add = glib_idle_add_wait if wait else glib_idle_add_once
    idle_add(_store_extend, store, things)
    # not that regardless of wait, None is returned
```